sdk/test/trace/tracer_test.cc

```diff
@@ -238,3 +241,125 @@ TEST(Tracer, SpanSetAttribute)
+  auto span_on_1 = tracer_on->StartSpan("span with AlwaysOn sampler 1");
+  auto span_on_2 = tracer_on->StartSpan("span with AlwaysOn sampler 2");
+
+  span_on_1->SetAttribute("attr1", 3.1);
+
+  span_on_2->End();
+  span_on_1->End();
+  ASSERT_EQ(2, spans_received_on->size());
+  auto &span_data_on_1 = spans_received_on->at(1);  // span 2 ends first.
+  auto &span_data_on_2 = spans_received_on->at(0);
+  ASSERT_EQ(3.1, nostd::get<double>(span_data_on_1->GetAttributes().at("attr1")));
+  ASSERT_LT(std::chrono::nanoseconds(0), span_data_on_1->GetStartTime().time_since_epoch());
+  ASSERT_LT(std::chrono::nanoseconds(0), span_data_on_1->GetDuration());
+  ASSERT_LT(std::chrono::nanoseconds(0), span_data_on_2->GetStartTime().time_since_epoch());
+  ASSERT_LT(std::chrono::nanoseconds(0), span_data_on_2->GetDuration());
+}
+
+TEST(Tracer, TestAlwaysOffSampler)
+{
+  std::shared_ptr<std::vector<std::unique_ptr<SpanData>>> spans_received_off(
+      new std::vector<std::unique_ptr<SpanData>>);
+  auto tracer_off = initTracer(spans_received_off, std::make_shared<AlwaysOffSampler>());
+  // Testing AlwaysOff sampler
+  auto span_off_1 = tracer_off->StartSpan("span with AlwaysOff sampler 1");
+  auto span_off_2 = tracer_off->StartSpan("span with AlwaysOff sampler 2");
+
+  span_off_1->SetAttribute("attr1", 3.1);  // Not recorded.
+
+  span_off_2->End();
+  span_off_1->End();
+
+  // The tracer export nothing with an AlwaysOff sampler
+  ASSERT_EQ(0, spans_received_off->size());
+}
+
+TEST(Tracer, TestParentOrElseSampler)
+{
+  std::shared_ptr<std::vector<std::unique_ptr<SpanData>>> spans_received_parent_on(
+      new std::vector<std::unique_ptr<SpanData>>);
+
+  // Current ShouldSample always pass an empty ParentContext,
+  // so this sampler will work as an AlwaysOnSampler.
+  auto tracer_parent_on =
+      initTracer(spans_received_parent_on,
+                 std::make_shared<ParentOrElseSampler>(std::make_shared<AlwaysOnSampler>()));
+
+  // Testing ParentOrElse sampler
+  auto span_parent_on_1 = tracer_parent_on->StartSpan("span with Parent-or-else sampler 1");
+  auto span_parent_on_2 = tracer_parent_on->StartSpan("span with Parent-or-else sampler 2");
+
+  span_parent_on_1->SetAttribute("attr1", 3.1);
+
+  span_parent_on_2->End();
+  span_parent_on_1->End();
+  ASSERT_EQ(2, spans_received_parent_on->size());
+  auto &span_data_parent_on_1 = spans_received_parent_on->at(1);
+  auto &span_data_parent_on_2 = spans_received_parent_on->at(0);
+  ASSERT_EQ(3.1, nostd::get<double>(span_data_parent_on_1->GetAttributes().at("attr1")));
+  ASSERT_LT(std::chrono::nanoseconds(0), span_data_parent_on_1->GetStartTime().time_since_epoch());
+  ASSERT_LT(std::chrono::nanoseconds(0), span_data_parent_on_1->GetDuration());
+  ASSERT_LT(std::chrono::nanoseconds(0), span_data_parent_on_2->GetStartTime().time_since_epoch());
+  ASSERT_LT(std::chrono::nanoseconds(0), span_data_parent_on_2->GetDuration());
+}
+
+TEST(Tracer, TestProbabilitySampler)
```

Comment:
I'm open to any other opinions here, but I'm in favor of leaving this test out for now. It doesn't to anything of use, as we don't have valid trace ids yet. And it will randomly break once we properly support trace id creation.

Reply:
Agreed.